Yoast/wordpress-seo
# Ensures that an indexable always has a permalink

## Context


* When the permalink structure changes (e.g. from numerical to slug) we set all of the permalinks of the indexables to `null`. However, we want the indexables to always have a permalink. So, whenever an indexable is retrieved from the database, we check if it has a permalink, if not: we generate it and save the indexable again with the permalink included.

## Summary


This PR can be summarized in the following changelog entry:

* Whenever an indexable is retrieved from the database, we check if it has a permalink. If not: we generate it and save the indexable with the permalink included.

## Relevant technical choices:

*

## Test instructions

This PR can be tested by following these steps:

(Copied from https://github.com/Yoast/bugreports/issues/962#issuecomment-612002431)

Install:

- Yoast SEO RC
- Yoast Test helper

### Steps to reproduce
1. Checkout `release/14.0`.
1. Enable the Books post type in Test Helper.  
2. Make sure breadcrumbs are enabled, and make sure the post type archive shows in the breadcrumbs for the books post type.
3. Create a New Book post, publish it.
4. Go to Settings -> Permalinks, switch to Numeric, hit save, switch back to Post name, hit save. 
   * This makes sure the permalink column in the indexables table is empty.
5. Go to your New Book post and view the source, the Post type archive has a section in the breadcrumbs schema, but it's `null`, because that permalink is not indexed yet.
6. Go to the post type archive in your browser.
7. Go back to your New Book post.
8. Now the breadcrumbs for the archive _should_ show properly.

### Steps to check if the bug has been fixed
1. Checkout this branch (`bugreports-962-lazy-load-indexable-hierarchy-ancestor`).
2. Go to Settings -> Permalinks, switch to Numeric, hit save, switch back to Post name, hit save. 
   * This makes sure the permalink column in the indexables table is empty.
3. Go to your Book post and view the source.
4. Check the breadcrumbs Schema, it should be correct.
    * E.g. every `ListItem` piece should have an `item` with a correct `@id` and `url` parameter.

### Test for categories.
1. Set the taxonomy to show for posts to 'Category' in the breadcrumbs settings 
   * Located in _Yoast SEO_ > _Search Appearance_ > _Breadcrumbs_ > _Taxonomy to show in breadcrumbs for content types_.
2. Follow the same steps as in 'reproduce' and 'check if fixed', but now for a single post that has a primary category.

### Test for tags.
1. Set the taxonomy to show for posts to 'Tag' in the breadcrumbs settings 
   * Located in _Yoast SEO_ > _Search Appearance_ > _Breadcrumbs_ > _Taxonomy to show in breadcrumbs for content types_.
2. Follow the same steps as in 'reproduce' and 'check if fixed', but now for a single post that has one or more tags.

## UI changes

## Documentation

## Quality assurance


Fixes #
